rules/opentitan_test.bzl

```diff
@@ -297,24 +300,43 @@ def opentitan_functest(
 
     deps = kwargs.pop("deps", [])
     all_tests = []
+    target_params = {}
 
     # Only build SW for devices we are running tests on.
     devices_to_build_for = []
     for target in targets:
         if target == "dv":
             devices_to_build_for.append("sim_dv")
+            target_params["sim_dv"] = dv_params() if not dv else dv
         elif target == "verilator":
             devices_to_build_for.append("sim_verilator")
-        elif target == "cw310":
+            target_params["sim_verilator"] = verilator_params() if not verilator else verilator
+        elif target in ["cw310_rom", "cw310_test_rom"]:
             devices_to_build_for.append("fpga_cw310")
+
+            # Deep-copy the `cw310` dict.
+            cw310_ = cw310_params() if not cw310 else {k: v for k, v in cw310.items()}
+            if "test_rom" in target:
+                cw310_["rom_kind"] = "testrom"
+                cw310_["bitstream"] = "@//hw/bitstream:test_rom"
+                cw310_["tags"].append("cw310_test_rom")
+                target_params["fpga_cw310_test_rom"] = cw310_
+            elif "rom" in target:
+                cw310_["rom_kind"] = "rom"
+                cw310_["bitstream"] = "@//hw/bitstream:rom"
+                cw310_["tags"].append("cw310_rom")
+                target_params["fpga_cw310_rom"] = cw310_
+            else:
+                fail("Expected to find `test_rom` or `rom` in the target name")
         else:
             fail("invalid target. Target must be in {}".format(VALID_TARGETS))
+    devices_to_build_for = collections.uniq(devices_to_build_for)
 
     # Handle the special case were the test is run at the ROM stage.
     if test_in_rom:
         if ot_flash_binary:
             fail("Tests that run in ROM stage cannot use pre-built flash binary.")
-        if "cw310" in targets:
+        if any(["cw310" in t for t in targets]):
```

Comment:
Nit: I still find this use of `in` for substring checking unintuitive/error-prone. At minimum, could we check for `"_cw310_"` substrings?

I feel like we're going to accidentally typo these strings and run into bugs. At minimum, we should probably define constants like `TARGET_FPGA_CW310_ROM = "fpga_cw310_rom"` and use the symbols instead of string literals.

Reply:
Sure, but this approach is widespread in this file and I don't think "_cw310_" is any safer for that matter.